src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+                        FSEntry::Dir(DirEntry {
+                            properties: Some(Mode(2, false)),
+                            children: [(
+                                "b",
+                                FSEntry::Dir(DirEntry {
+                                    properties: Some(Mode(1, true)),
+                                    children: [
+                                        ("c", FSEntry::File(Mode(3, false)).into()),
+                                        (
+                                            "f",
+                                            FSEntry::Dir(DirEntry {
+                                                properties: None,
+                                                children: [(
+                                                    "g",
+                                                    FSEntry::File(Mode(6, false)).into()
+                                                )]
+                                                .into_iter()
+                                                .collect(),
+                                            })
+                                            .into()
+                                        ),
+                                    ]
+                                    .into_iter()
+                                    .collect(),
+                                })
+                                .into()
+                            )]
+                            .into_iter()
+                            .collect(),
+                        })
+                        .into()
+                    ),
+                    (
+                        "d",
+                        FSEntry::Dir(DirEntry {
+                            properties: Some(Mode(4, true)),
+                            children: BTreeMap::new(),
+                        })
+                        .into()
+                    ),
+                    ("e", FSEntry::File(Mode(5, false)).into())
+                ]
+                .into_iter()
+                .collect()
+            );
+        }
+    }
+}
+
+pub mod handle_creation {
+    use displaydoc::Display;
+    use thiserror::Error;
+
+    use std::cmp;
+    use std::collections::{HashMap, VecDeque};
+    use std::fs;
+    use std::hash;
+    use std::io;
+    use std::path::{Path, PathBuf};
+
+    use crate::types::ZipFileData;
+
+    use super::path_splitting::{DirEntry, FSEntry};
+
+    /// Errors encountered when creating output handles for extracting entries to.
+    #[derive(Debug, Display, Error)]
+    pub enum HandleCreationError {
+        /// i/o error: {0}
+        Io(#[from] io::Error),
+    }
+
+    /// Wrapper for memory location of the ZipFileData in Shared.
+    ///
+    /// Enables quick comparison and hash table lookup without needing to implement Hash/Eq for
+    /// ZipFileData more generally.
+    #[derive(Debug)]
+    pub(crate) struct ZipDataHandle<'a>(&'a ZipFileData);
```

Comment:
Use [ByAddress](https://docs.rs/by_address/latest/by_address/struct.ByAddress.html) rather than reinventing the wheel. Also, if I understand correctly, this pointer needs to be pinned if the hash is to be reliably stable (i.e. you need to use `ByAddress<Pin<&'a ZipFileData>>`.

Reply:
I think you're right about the pinning part, thanks so much! I wasn't aware of `ByAddress` -- looking into that now!

Reply:
Done!!